File: recyclerview-swipe/src/main/java/com/yanzhenjie/recyclerview/swipe/widget/StickyNestedScrollView.java
/*
 * Copyright 2017 Yan Zhenjie
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yanzhenjie.recyclerview.swipe.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.v4.widget.NestedScrollView;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * <p>https://github.com/yanzhenjie/StickyScrollView</p>
 * Created by YanZhenjie on 2017/7/20.
 */
public class StickyNestedScrollView extends NestedScrollView {

    public interface OnViewStickyListener {

        void onSticky(View view);

        void onUnSticky(View view);
    }

    /**
     * Tag for views that should stick and have constant drawing. e.g. TextViews, ImageViews etc
     */
    public static final String STICKY_TAG = "sticky";
    /**
     * Flag for views that should stick and have non-constant drawing. e.g. Buttons, ProgressBars etc
     */
    public static final String FLAG_NONCONSTANT = "-nonconstant";
    /**
     * Flag for views that have aren't fully opaque
     */
    public static final String FLAG_HASTRANSPARENCY = "-hastransparency";
    /**
     * Default height of the shadow peeking out below the stuck view.
     */
    private static final int DEFAULT_SHADOW_HEIGHT = 10; // dp;
    private ArrayList<View> stickyViews;
    private View currentlyStickingView;
    private float stickyViewTopOffset;
    private final Runnable invalidateRunnable = new Runnable() {
        @Override
        public void run() {
            if (currentlyStickingView != null) {
                int l = getLeftForViewRelativeOnlyChild(currentlyStickingView);
                int t = getBottomForViewRelativeOnlyChild(currentlyStickingView);
                int r = getRightForViewRelativeOnlyChild(currentlyStickingView);
                int b = (int) (getScrollY() + (currentlyStickingView.getHeight() + stickyViewTopOffset));
                invalidate(l, t, r, b);
            }
            postDelayed(this, 16);
        }
    };
    private int stickyViewLeftOffset;
    private boolean redirectTouchesToStickyView;
    private boolean clippingToPadding;
    private boolean clipToPaddingHasBeenSet;
    private int mShadowHeight = DEFAULT_SHADOW_HEIGHT;
    private Drawable mShadowDrawable;
    private boolean hasNotDoneActionDown = true;

    private List<OnViewStickyListener> mOnViewStickyListeners;

    public StickyNestedScrollView(Context context) {
        this(context, null);
    }

    public StickyNestedScrollView(Context context, AttributeSet attrs) {
        this(context, attrs, android.R.attr.scrollViewStyle);
    }

    public StickyNestedScrollView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setup();
    }

    public void addOnViewStickyListener(OnViewStickyListener stickyListener) {
        if (mOnViewStickyListeners == null)
            mOnViewStickyListeners = new ArrayList<>();
        mOnViewStickyListeners.add(stickyListener);
    }

    public void removeOnViewStickyListener(OnViewStickyListener stickyListener) {
        if (mOnViewStickyListeners != null)
            mOnViewStickyListeners.remove(stickyListener);
    }

    public void clearOnViewStickyListener() {
        if (mOnViewStickyListeners != null)
            mOnViewStickyListeners.clear();
    }

    public void setShadowHeight(int height) {
        mShadowHeight = height;
    }

    public void setShadowDrawable(Drawable shadowDrawable) {
        mShadowDrawable = shadowDrawable;
    }

    public void setup() {
        stickyViews = new ArrayList<>();
    }

    private int getLeftForViewRelativeOnlyChild(View v) {
        int left = v.getLeft();
        while (v.getParent() != null && v.getParent() != getChildAt(0)) {
            v = (View) v.getParent();
            left += v.getLeft();
        }
        return left;
    }

    private int getTopForViewRelativeOnlyChild(View v) {
        int top = v.getTop();
        while (v.getParent() != null && v.getParent() != getChildAt(0)) {
            v = (View) v.getParent();
            top += v.getTop();
        }
        return top;
    }

    private int getRightForViewRelativeOnlyChild(View v) {
        int right = v.getRight();
        while (v.getParent() != null && v.getParent() != getChildAt(0)) {
            v = (View) v.getParent();
            right += v.getRight();
        }
        return right;
    }

    private int getBottomForViewRelativeOnlyChild(View v) {
        int bottom = v.getBottom();
        while (v.getParent() != null && v.getParent() != getChildAt(0)) {
            v = (View) v.getParent();
            bottom += v.getBottom();
        }
        return bottom;
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        if (!clipToPaddingHasBeenSet) {
            clippingToPadding = true;
        }
        notifyHierarchyChanged();
    }

    @Override
    public void setClipToPadding(boolean clipToPadding) {
        super.setClipToPadding(clipToPadding);
        clippingToPadding = clipToPadding;
        clipToPaddingHasBeenSet = true;
    }

    @Override
    public void addView(View child) {
        super.addView(child);
        findStickyViews(child);
    }

    @Override
    public void addView(View child, int index) {
        super.addView(child, index);
        findStickyViews(child);
    }

    @Override
    public void addView(View child, int index, ViewGroup.LayoutParams params) {
        super.addView(child, index, params);
        findStickyViews(child);
    }

    @Override
    public void addView(View child, int width, int height) {
        super.addView(child, width, height);
        findStickyViews(child);
    }

    @Override
    public void addView(View child, ViewGroup.LayoutParams params) {
        super.addView(child, params);
        findStickyViews(child);
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        super.dispatchDraw(canvas);
        if (currentlyStickingView != null) {
            canvas.save();
            canvas.translate(getPaddingLeft() + stickyViewLeftOffset, getScrollY()
                    + stickyViewTopOffset + (clippingToPadding ? getPaddingTop() : 0));
            canvas.clipRect(0, (clippingToPadding ? -stickyViewTopOffset : 0),
                    getWidth() - stickyViewLeftOffset,
                    currentlyStickingView.getHeight() + mShadowHeight + 1);
            if (mShadowDrawable != null) {
                int left = 0;
                int top = currentlyStickingView.getHeight();
                int right = currentlyStickingView.getWidth();
                int bottom = currentlyStickingView.getHeight() + mShadowHeight;
                mShadowDrawable.setBounds(left, top, right, bottom);
                mShadowDrawable.draw(canvas);
            }
            canvas.clipRect(0, (clippingToPadding ? -stickyViewTopOffset : 0), getWidth(),
                    currentlyStickingView.getHeight());
            if (getStringTagForView(currentlyStickingView).contains(FLAG_HASTRANSPARENCY)) {
                showView(currentlyStickingView);
                currentlyStickingView.draw(canvas);
                hideView(currentlyStickingView);
            } else {
                currentlyStickingView.draw(canvas);
            }
            canvas.restore();
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (ev.getAction() == MotionEvent.ACTION_DOWN) {
            redirectTouchesToStickyView = true;
        }
        if (redirectTouchesToStickyView) {
            redirectTouchesToStickyView = currentlyStickingView != null;
            if (redirectTouchesToStickyView) {
                redirectTouchesToStickyView =
                        ev.getY() <= (currentlyStickingView.getHeight() + stickyViewTopOffset) &&
                                ev.getX() >= getLeftForViewRelativeOnlyChild(currentlyStickingView) &&
                                ev.getX() <= getRightForViewRelativeOnlyChild(currentlyStickingView);
            }
        } else if (currentlyStickingView == null) {
            redirectTouchesToStickyView = false;
        }
        if (redirectTouchesToStickyView) {
            ev.offsetLocation(0, -1 * ((getScrollY() + stickyViewTopOffset)
                    - getTopForViewRelativeOnlyChild(currentlyStickingView)));
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        if (redirectTouchesToStickyView) {
            ev.offsetLocation(0, ((getScrollY() + stickyViewTopOffset)
                    - getTopForViewRelativeOnlyChild(currentlyStickingView)));
        }
        if (ev.getAction() == MotionEvent.ACTION_DOWN) {
            hasNotDoneActionDown = false;
        }
        if (hasNotDoneActionDown) {
            MotionEvent down = MotionEvent.obtain(ev);
            down.setAction(MotionEvent.ACTION_DOWN);
            super.onTouchEvent(down);
            hasNotDoneActionDown = false;
        }
        if (ev.getAction() == MotionEvent.ACTION_UP || ev.getAction() == MotionEvent.ACTION_CANCEL) {
            hasNotDoneActionDown = true;
        }
        return super.onTouchEvent(ev);
    }

    @Override
    protected void onScrollChanged(int l, int t, int oldl, int oldt) {
        super.onScrollChanged(l, t, oldl, oldt);
        doTheStickyThing();
    }

    private void doTheStickyThing() {
        View viewThatShouldStick = null;
        View approachingView = null;
        for (View v : stickyViews) {
            int viewTop = getTopForViewRelativeOnlyChild(v) - getScrollY()
                    + (clippingToPadding ? 0 : getPaddingTop());
            if (viewTop <= 0) {
                if (viewThatShouldStick == null || viewTop >
                        (getTopForViewRelativeOnlyChild(viewThatShouldStick)
                                - getScrollY() + (clippingToPadding ? 0 : getPaddingTop()))) {
                    viewThatShouldStick = v;
                }
            } else {
                if (approachingView == null || viewTop < (getTopForViewRelativeOnlyChild(approachingView)
                        - getScrollY() + (clippingToPadding ? 0 : getPaddingTop()))) {
                    approachingView = v;
                }
            }
        }
        if (viewThatShouldStick != null) {
            stickyViewTopOffset = approachingView == null ? 0 :
                    Math.min(0, getTopForViewRelativeOnlyChild(approachingView) - getScrollY()
                            + (clippingToPadding ? 0 : getPaddingTop()) - viewThatShouldStick.getHeight());
            if (viewThatShouldStick != currentlyStickingView) {
                if (currentlyStickingView != null) {
                    if (mOnViewStickyListeners != null)
                        for (OnViewStickyListener onViewStickyListener : mOnViewStickyListeners)
                            onViewStickyListener.onUnSticky(currentlyStickingView);
                    stopStickingCurrentlyStickingView();
                }
                // only compute the left offset when we start sticking.
                stickyViewLeftOffset = getLeftForViewRelativeOnlyChild(viewThatShouldStick);
                startStickingView(viewThatShouldStick);
                if (mOnViewStickyListeners != null)
                    for (OnViewStickyListener onViewStickyListener : mOnViewStickyListeners)
                        onViewStickyListener.onSticky(currentlyStickingView);
            }
        } else if (currentlyStickingView != null) {
            if (mOnViewStickyListeners != null)
                for (OnViewStickyListener onViewStickyListener : mOnViewStickyListeners)
                    onViewStickyListener.onUnSticky(currentlyStickingView);
            stopStickingCurrentlyStickingView();
        }
    }

    private void startStickingView(View viewThatShouldStick) {
        currentlyStickingView = viewThatShouldStick;
        if (currentlyStickingView != null) {
            if (getStringTagForView(currentlyStickingView).contains(FLAG_HASTRANSPARENCY)) {
                hideView(currentlyStickingView);
            }
            if (getStringTagForView(currentlyStickingView).contains(FLAG_NONCONSTANT)) {
                post(invalidateRunnable);
            }
        }
    }

    private void stopStickingCurrentlyStickingView() {
        if (getStringTagForView(currentlyStickingView).contains(FLAG_HASTRANSPARENCY)) {
            showView(currentlyStickingView);
        }
        currentlyStickingView = null;
        removeCallbacks(invalidateRunnable);
    }

    @Override
    protected void onDetachedFromWindow() {
        removeCallbacks(invalidateRunnable);
        super.onDetachedFromWindow();
    }

    /**
     * Notify that the sticky attribute has been added or removed from one or more views in the View hierarchy
     */
    public void notifyStickyAttributeChanged() {
        notifyHierarchyChanged();
    }

    private void notifyHierarchyChanged() {
        if (currentlyStickingView != null) {
            stopStickingCurrentlyStickingView();
        }
        stickyViews.clear();
        findStickyViews(getChildAt(0));
        doTheStickyThing();
        invalidate();
    }

    private void findStickyViews(View v) {
        if (!detainStickyView(v) && (v instanceof ViewGroup)) {
            ViewGroup vg = (ViewGroup) v;
            for (int i = 0; i < vg.getChildCount(); i++)
                findStickyViews(vg.getChildAt(i));
        }
    }

    private boolean detainStickyView(View view) {
        String tag = getStringTagForView(view);
        if (tag.contains(STICKY_TAG)) {
            stickyViews.add(view);
            return true;
        }
        return false;
    }

    private String getStringTagForView(View v) {
        Object tagObject = v.getTag();
        return String.valueOf(tagObject);
    }

    private void hideView(View v) {
        v.setAlpha(0);
    }

    private void showView(View v) {
        v.setAlpha(1);
    }

}

File: recyclerview-swipe/src/main/java/com/yanzhenjie/recyclerview/swipe/widget/StickyNestedScrollView.kt
/*
 * Copyright 2017 Yan Zhenjie
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yanzhenjie.recyclerview.swipe.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.support.v4.widget.NestedScrollView
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

/**
 *
 * https://github.com/yanzhenjie/StickyScrollView
 * Created by YanZhenjie on 2017/7/20.
 */
class StickyNestedScrollView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = android.R.attr.scrollViewStyle) : NestedScrollView(context, attrs, defStyle) {

  private var stickyViews: ArrayList<View>? = null
  private var currentlyStickingView: View? = null
  private var stickyViewTopOffset: Float = 0.toFloat()
  private val invalidateRunnable = object : Runnable {
    override fun run() {
      if (currentlyStickingView != null) {
        val l = getLeftForViewRelativeOnlyChild(currentlyStickingView)
        val t = getBottomForViewRelativeOnlyChild(currentlyStickingView)
        val r = getRightForViewRelativeOnlyChild(currentlyStickingView)
        val b = (scrollY + (currentlyStickingView!!.height + stickyViewTopOffset)).toInt()
        invalidate(l, t, r, b)
      }
      postDelayed(this, 16)
    }
  }
  private var stickyViewLeftOffset: Int = 0
  private var redirectTouchesToStickyView: Boolean = false
  private var clippingToPadding: Boolean = false
  private var clipToPaddingHasBeenSet: Boolean = false
  private var mShadowHeight = DEFAULT_SHADOW_HEIGHT
  private var mShadowDrawable: Drawable? = null
  private var hasNotDoneActionDown = true

  private var mOnViewStickyListeners: MutableList<OnViewStickyListener>? = null

  interface OnViewStickyListener {
    fun onSticky(view: View?)
    fun onUnSticky(view: View)
  }

  init {
    setup()
  }

  fun addOnViewStickyListener(stickyListener: OnViewStickyListener) {
    var listeners = mOnViewStickyListeners

    if (listeners == null) {
      listeners = ArrayList()
      mOnViewStickyListeners = listeners
    }
    listeners.add(stickyListener)
  }

  fun removeOnViewStickyListener(stickyListener: OnViewStickyListener) {
    mOnViewStickyListeners?.remove(stickyListener)
  }

  fun clearOnViewStickyListener() {
    mOnViewStickyListeners?.clear()
  }

  fun setShadowHeight(height: Int) {
    mShadowHeight = height
  }

  fun setShadowDrawable(shadowDrawable: Drawable) {
    mShadowDrawable = shadowDrawable
  }

  fun setup() {
    stickyViews = ArrayList()
  }

  private fun getLeftForViewRelativeOnlyChild(view: View?) = getBoundaryOffsetForViewRelativeOnlyChild(view, { it.left })
  private fun getTopForViewRelativeOnlyChild(view: View?) = getBoundaryOffsetForViewRelativeOnlyChild(view, { it.top })
  private fun getRightForViewRelativeOnlyChild(view: View?) = getBoundaryOffsetForViewRelativeOnlyChild(view, { it.right })
  private fun getBottomForViewRelativeOnlyChild(view: View?) = getBoundaryOffsetForViewRelativeOnlyChild(view, { it.bottom })

  /**
   * Returns boundary offset for view relative only child.
   *
   * @param view view.
   * @param getBoundaryOffset get boundary offset: fun(view): boundaryOffset.
   */
  private fun getBoundaryOffsetForViewRelativeOnlyChild(view: View?, getBoundaryOffset: (View) -> Int): Int {
    var childView = view ?: return 0
    var childParent = childView.parent
    var childLeft = getBoundaryOffset(childView)

    while (childParent != null && childParent != getChildAt(0) && childParent is View) {
      childView = childParent
      childParent = childView.parent
      childLeft += getBoundaryOffset(childView)
    }
    return childLeft
  }

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
    super.onLayout(changed, l, t, r, b)
    if (!clipToPaddingHasBeenSet) {
      clippingToPadding = true
    }
    notifyHierarchyChanged()
  }

  override fun setClipToPadding(clipToPadding: Boolean) {
    super.setClipToPadding(clipToPadding)
    clippingToPadding = clipToPadding
    clipToPaddingHasBeenSet = true
  }

  override fun addView(child: View) {
    super.addView(child)
    findStickyViews(child)
  }

  override fun addView(child: View, index: Int) {
    super.addView(child, index)
    findStickyViews(child)
  }

  override fun addView(child: View, index: Int, params: ViewGroup.LayoutParams) {
    super.addView(child, index, params)
    findStickyViews(child)
  }

  override fun addView(child: View, width: Int, height: Int) {
    super.addView(child, width, height)
    findStickyViews(child)
  }

  override fun addView(child: View, params: ViewGroup.LayoutParams) {
    super.addView(child, params)
    findStickyViews(child)
  }

  override fun dispatchDraw(canvas: Canvas) {
    super.dispatchDraw(canvas)
    val stickingView = currentlyStickingView ?: return
    val shadowDrawable = mShadowDrawable

    canvas.save()
    canvas.translate(
        (paddingLeft + stickyViewLeftOffset).toFloat(),
        scrollY.toFloat()+ stickyViewTopOffset + (if (clippingToPadding) paddingTop else 0).toFloat())

    canvas.clipRect(
        0,
        if (clippingToPadding) -stickyViewTopOffset.toInt() else 0,
        width - stickyViewLeftOffset,
        stickingView.height + mShadowHeight + 1
    )
    if (shadowDrawable != null) {
      val left = 0
      val top = stickingView.height
      val right = stickingView.width
      val bottom = stickingView.height + mShadowHeight

      shadowDrawable.setBounds(left, top, right, bottom)
      shadowDrawable.draw(canvas)
    }

    canvas.clipRect(
        0,
        if (clippingToPadding) -stickyViewTopOffset.toInt() else 0,
        width,
        stickingView.height
    )
    if (getStringTagForView(stickingView).contains(FLAG_HASTRANSPARENCY)) {
      showView(stickingView)
      stickingView.draw(canvas)
      hideView(stickingView)
    } else {
      stickingView.draw(canvas)
    }
    canvas.restore()
  }

  override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
    val stickingView = currentlyStickingView

    if (ev.action == MotionEvent.ACTION_DOWN) {
      redirectTouchesToStickyView = true
    }
    if (redirectTouchesToStickyView) {
      redirectTouchesToStickyView = currentlyStickingView != null

      if (redirectTouchesToStickyView) {
        redirectTouchesToStickyView = stickingView != null && ev.y <= stickingView.height + stickyViewTopOffset &&
            ev.x >= getLeftForViewRelativeOnlyChild(currentlyStickingView) &&
            ev.x <= getRightForViewRelativeOnlyChild(currentlyStickingView)
      }
    } else if (currentlyStickingView == null) {
      redirectTouchesToStickyView = false
    }
    if (redirectTouchesToStickyView) {
      ev.offsetLocation(0f, -1 * (scrollY + stickyViewTopOffset - getTopForViewRelativeOnlyChild(currentlyStickingView)))
    }
    return super.dispatchTouchEvent(ev)
  }

  override fun onTouchEvent(ev: MotionEvent): Boolean {
    if (redirectTouchesToStickyView) {
      ev.offsetLocation(0f, scrollY + stickyViewTopOffset - getTopForViewRelativeOnlyChild(currentlyStickingView))
    }
    if (ev.action == MotionEvent.ACTION_DOWN) {
      hasNotDoneActionDown = false
    }
    if (hasNotDoneActionDown) {
      val down = MotionEvent.obtain(ev)
      down.action = MotionEvent.ACTION_DOWN
      super.onTouchEvent(down)
      hasNotDoneActionDown = false
    }
    if (ev.action == MotionEvent.ACTION_UP || ev.action == MotionEvent.ACTION_CANCEL) {
      hasNotDoneActionDown = true
    }
    return super.onTouchEvent(ev)
  }

  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
    super.onScrollChanged(l, t, oldl, oldt)
    doTheStickyThing()
  }

  private fun doTheStickyThing() {
    var viewThatShouldStick: View? = null
    var approachingView: View? = null

    stickyViews?.forEach {
      val viewTop = getTopForViewRelativeOnlyChild(it) - scrollY + if (clippingToPadding) 0 else paddingTop

      if (viewTop <= 0) {
        if (viewThatShouldStick == null || viewTop > getTopForViewRelativeOnlyChild(viewThatShouldStick) - scrollY + if (clippingToPadding) 0 else paddingTop) {
          viewThatShouldStick = it
        }
      } else {
        if (approachingView == null || viewTop < getTopForViewRelativeOnlyChild(approachingView) - scrollY + if (clippingToPadding) 0 else paddingTop) {
          approachingView = it
        }
      }
    }
    val oldStickingView = currentlyStickingView
    val newStickingView = viewThatShouldStick

    if (newStickingView != null) {
      stickyViewTopOffset = if (approachingView == null) {
        0F
      } else {
        Math.min(0, getTopForViewRelativeOnlyChild(approachingView) - scrollY + if (clippingToPadding) 0 else paddingTop - newStickingView.height).toFloat()
      }

      if (newStickingView !== oldStickingView) {
        if (oldStickingView != null) {
          mOnViewStickyListeners?.forEach { it.onUnSticky(oldStickingView) }
          stopStickingCurrentlyStickingView()
        }
        // only compute the left offset when we start sticking.
        stickyViewLeftOffset = getLeftForViewRelativeOnlyChild(newStickingView)
        startStickingView(newStickingView)

        mOnViewStickyListeners?.forEach { it.onSticky(newStickingView) }
      }
    } else if (oldStickingView != null) {
      mOnViewStickyListeners?.forEach { it.onUnSticky(oldStickingView) }
      stopStickingCurrentlyStickingView()
    }
  }

  private fun startStickingView(viewThatShouldStick: View?) {
    currentlyStickingView = viewThatShouldStick

    if (viewThatShouldStick != null) {
      if (getStringTagForView(viewThatShouldStick).contains(FLAG_HASTRANSPARENCY)) {
        hideView(viewThatShouldStick)
      }
      if (getStringTagForView(viewThatShouldStick).contains(FLAG_NONCONSTANT)) {
        post(invalidateRunnable)
      }
    }
  }

  private fun stopStickingCurrentlyStickingView() {
    if (getStringTagForView(currentlyStickingView).contains(FLAG_HASTRANSPARENCY)) {
      showView(currentlyStickingView)
    }
    currentlyStickingView = null
    removeCallbacks(invalidateRunnable)
  }

  override fun onDetachedFromWindow() {
    removeCallbacks(invalidateRunnable)
    super.onDetachedFromWindow()
  }

  /**
   * Notify that the sticky attribute has been added or removed from one or more views in the View hierarchy
   */
  fun notifyStickyAttributeChanged() {
    notifyHierarchyChanged()
  }

  private fun notifyHierarchyChanged() {
    if (currentlyStickingView != null) {
      stopStickingCurrentlyStickingView()
    }
    stickyViews?.clear()
    findStickyViews(getChildAt(0))
    doTheStickyThing()
    invalidate()
  }

  private fun findStickyViews(v: View) {
    if (!detainStickyView(v) && v is ViewGroup) {
      for (i in 0 until v.childCount) {
        findStickyViews(v.getChildAt(i))
      }
    }
  }

  private fun detainStickyView(view: View): Boolean {
    val tag = getStringTagForView(view)

    if (tag.contains(STICKY_TAG)) {
      stickyViews?.add(view)
      return true
    }
    return false
  }

  private fun getStringTagForView(v: View?): String {
    val tagObject = v?.tag ?: return ""
    return tagObject.toString()
  }

  private fun hideView(v: View) {
    v.alpha = 0f
  }

  private fun showView(v: View?) {
    v?.alpha = 1f
  }

  companion object {

    /**
     * Tag for views that should stick and have constant drawing. e.g. TextViews, ImageViews etc
     */
    const val STICKY_TAG = "sticky"

    /**
     * Flag for views that should stick and have non-constant drawing. e.g. Buttons, ProgressBars etc
     */
    const val FLAG_NONCONSTANT = "-nonconstant"

    /**
     * Flag for views that have aren't fully opaque
     */
    const val FLAG_HASTRANSPARENCY = "-hastransparency"

    /**
     * Default height of the shadow peeking out below the stuck view.
     */
    private const val DEFAULT_SHADOW_HEIGHT = 10 // dp;
  }

}